File: src/lib/url-metadata.ts
export interface UrlMetadata {
  title: string | null;
  favicon: string | null;
  fetchedAt: number;
}

function isAllowedUrl(url: string): boolean {
  try {
    const parsed = new URL(url);

    if (!["http:", "https:"].includes(parsed.protocol)) {
      return false;
    }

    const hostname = parsed.hostname.toLowerCase();

    if (
      hostname === "localhost" ||
      hostname === "127.0.0.1" ||
      hostname === "::1" ||
      hostname === "0.0.0.0"
    ) {
      return false;
    }

    const ipMatch = hostname.match(/^(\d+)\.(\d+)\.(\d+)\.(\d+)$/);
    if (ipMatch) {
      const [, a, b] = ipMatch.map(Number);
      if (a === 10) return false;
      if (a === 172 && b >= 16 && b <= 31) return false;
      if (a === 192 && b === 168) return false;
      if (a === 169 && b === 254) return false;
      if (a === 127) return false;
    }

    return true;
  } catch {
    return false;
  }
}

interface FetchResult {
  metadata: UrlMetadata;
  success: boolean;
}

async function fetchUrlMetadata(url: string): Promise<FetchResult> {
  if (!isAllowedUrl(url)) {
    return {
      metadata: { title: null, favicon: null, fetchedAt: Date.now() },
      success: false,
    };
  }

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 10000);

    const response = await fetch(url, {
      signal: controller.signal,
      headers: {
        "User-Agent": "Mozilla/5.0 (compatible; minimal/1.0)",
        Accept: "text/html,application/xhtml+xml",
      },
    });

    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const html = await response.text();
    const metadata = parseHtmlMetadata(html, url);

    return {
      metadata: { ...metadata, fetchedAt: Date.now() },
      success: true,
    };
  } catch {
    return {
      metadata: { title: null, favicon: null, fetchedAt: Date.now() },
      success: false,
    };
  }
}

function parseHtmlMetadata(
  html: string,
  baseUrl: string
): Omit<UrlMetadata, "fetchedAt"> {
  let title: string | null = null;
  let favicon: string | null = null;

  const ogTitleMatch = html.match(
    /<meta[^>]*property=["']og:title["'][^>]*content=["']([^"']+)["']/i
  );
  if (ogTitleMatch) {
    title = decodeHtmlEntities(ogTitleMatch[1]);
  } else {
    const ogTitleMatchAlt = html.match(
      /<meta[^>]*content=["']([^"']+)["'][^>]*property=["']og:title["']/i
    );
    if (ogTitleMatchAlt) {
      title = decodeHtmlEntities(ogTitleMatchAlt[1]);
    } else {
      const titleMatch = html.match(/<title[^>]*>([^<]+)<\/title>/i);
      if (titleMatch) {
        title = decodeHtmlEntities(titleMatch[1].trim());
      }
    }
  }

  const iconPatterns = [
    /<link[^>]*rel=["']apple-touch-icon["'][^>]*href=["']([^"']+)["']/i,
    /<link[^>]*href=["']([^"']+)["'][^>]*rel=["']apple-touch-icon["']/i,
    /<link[^>]*rel=["']icon["'][^>]*href=["']([^"']+)["']/i,
    /<link[^>]*href=["']([^"']+)["'][^>]*rel=["']icon["']/i,
    /<link[^>]*rel=["']shortcut icon["'][^>]*href=["']([^"']+)["']/i,
    /<link[^>]*href=["']([^"']+)["'][^>]*rel=["']shortcut icon["']/i,
  ];

  for (const pattern of iconPatterns) {
    const match = html.match(pattern);
    if (match) {
      const resolved = resolveUrl(match[1], baseUrl);
      if (resolved) {
        favicon = resolved;
        break;
      }
    }
  }

  if (!favicon) {
    try {
      const parsed = new URL(baseUrl);
      favicon = `${parsed.protocol}//${parsed.host}/favicon.ico`;
    } catch {}
  }

  return { title, favicon };
}

function resolveUrl(href: string, baseUrl: string): string | null {
  try {
    return new URL(href, baseUrl).href;
  } catch {
    return null;
  }
}

function decodeHtmlEntities(text: string): string {
  return text
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&#x27;/g, "'")
    .replace(/&#x2F;/g, "/")
    .replace(/&nbsp;/g, " ");
}

export async function getUrlMetadata(url: string): Promise<UrlMetadata> {
  const { metadata } = await fetchUrlMetadata(url);
  return metadata;
}

File: src-tauri/src/services/url_metadata.rs
use reqwest;
use scraper::{Html, Selector};

#[derive(Debug, Clone)]
pub struct UrlMetadata {
    pub title: Option<String>,
    pub favicon: Option<String>,
}

/// Fetch metadata (title and favicon) from a URL
pub async fn fetch_url_metadata(url: &str) -> Result<UrlMetadata, Box<dyn std::error::Error>> {
    // Validate URL
    let parsed_url = url::Url::parse(url)?;

    // Security: only allow http and https
    if parsed_url.scheme() != "http" && parsed_url.scheme() != "https" {
        return Err("Only HTTP and HTTPS URLs are allowed".into());
    }

    // Security: block localhost and private IP ranges
    if let Some(host) = parsed_url.host_str() {
        let host_lower = host.to_lowercase();
        if host_lower == "localhost"
            || host_lower == "127.0.0.1"
            || host_lower == "::1"
            || host_lower == "0.0.0.0"
            || is_private_ip(&host_lower)
        {
            return Err("Private IP addresses and localhost are not allowed".into());
        }
    }

    // Fetch the HTML with timeout
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .user_agent("Mozilla/5.0 (compatible; usedraft/1.0)")
        .build()?;

    let response = client.get(url).send().await?;

    if !response.status().is_success() {
        return Err(format!("HTTP error: {}", response.status()).into());
    }

    let html = response.text().await?;
    let document = Html::parse_document(&html);

    // Extract title
    let title = extract_title(&document);

    // Extract favicon
    let favicon = extract_favicon(&document, &parsed_url);

    Ok(UrlMetadata { title, favicon })
}

fn extract_title(document: &Html) -> Option<String> {
    // Try og:title first
    if let Ok(selector) = Selector::parse("meta[property='og:title']") {
        if let Some(element) = document.select(&selector).next() {
            if let Some(content) = element.value().attr("content") {
                return Some(decode_html_entities(content));
            }
        }
    }

    // Fallback to <title> tag
    if let Ok(selector) = Selector::parse("title") {
        if let Some(element) = document.select(&selector).next() {
            let title = element.text().collect::<String>().trim().to_string();
            if !title.is_empty() {
                return Some(decode_html_entities(&title));
            }
        }
    }

    None
}

fn extract_favicon(document: &Html, base_url: &url::Url) -> Option<String> {
    // Try different favicon link patterns
    let patterns = vec![
        "link[rel='apple-touch-icon']",
        "link[rel='icon']",
        "link[rel='shortcut icon']",
    ];

    for pattern in patterns {
        if let Ok(selector) = Selector::parse(pattern) {
            if let Some(element) = document.select(&selector).next() {
                if let Some(href) = element.value().attr("href") {
                    // Resolve relative URLs
                    if let Ok(favicon_url) = base_url.join(href) {
                        return Some(favicon_url.to_string());
                    }
                }
            }
        }
    }

    // Fallback to /favicon.ico
    if let Ok(favicon_url) = base_url.join("/favicon.ico") {
        return Some(favicon_url.to_string());
    }

    None
}

fn decode_html_entities(text: &str) -> String {
    text.replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace("&#39;", "'")
        .replace("&#x27;", "'")
        .replace("&#x2F;", "/")
        .replace("&nbsp;", " ")
}

fn is_private_ip(host: &str) -> bool {
    // Check for private IP ranges
    if let Ok(addr) = host.parse::<std::net::IpAddr>() {
        match addr {
            std::net::IpAddr::V4(ipv4) => {
                let octets = ipv4.octets();
                // 10.0.0.0/8
                if octets[0] == 10 {
                    return true;
                }
                // 172.16.0.0/12
                if octets[0] == 172 && (octets[1] >= 16 && octets[1] <= 31) {
                    return true;
                }
                // 192.168.0.0/16
                if octets[0] == 192 && octets[1] == 168 {
                    return true;
                }
                // 169.254.0.0/16 (link-local)
                if octets[0] == 169 && octets[1] == 254 {
                    return true;
                }
                // 127.0.0.0/8 (loopback)
                if octets[0] == 127 {
                    return true;
                }
            }
            std::net::IpAddr::V6(ipv6) => {
                // Check for IPv6 loopback and link-local
                if ipv6.is_loopback() {
                    return true;
                }
                // fe80::/10 link-local
                let segments = ipv6.segments();
                if (segments[0] & 0xffc0) == 0xfe80 {
                    return true;
                }
            }
        }
    }
    false
}
